qcodes/dataset/dependencies.py

```diff
@@ -1,6 +1,526 @@
+
+        self.dependencies: ParamSpecTree = dependencies
+        self.inferences: ParamSpecTree = inferences
+        self.standalones: FrozenSet[ParamSpecBase] = frozenset(standalones)
+
+        # The object is now complete, but for convenience, we form some more
+        # private attributes for easy look-up
+
+        self._id_to_paramspec: Dict[str, ParamSpecBase] = {}
+        for tree in (self.dependencies, self.inferences):
+            for ps, ps_tup in tree.items():
+                self._id_to_paramspec.update({self._id(ps): ps})
+                self._id_to_paramspec.update({self._id(pst): pst
+                                              for pst in ps_tup})
+        for ps in self.standalones:
+            self._id_to_paramspec.update({self._id(ps): ps})
+        self._paramspec_to_id = {v: k for k, v in self._id_to_paramspec.items()}
+
+        self._dependencies_inv: ParamSpecTree = self._invert_tree(
+            self.dependencies)
+        self._inferences_inv: ParamSpecTree = self._invert_tree(
+            self.inferences)
+
+        # Set operations are ~2x (or more) faster on strings than on hashable
+        # ParamSpecBase objects, hence the need for use of the following
+        # representation
+        self._deps_names: ParamNameTree = self._tree_of_names(
+            self.dependencies)
+        self._infs_names: ParamNameTree = self._tree_of_names(
+            self.inferences)
+
+    @staticmethod
+    def _tree_of_names(tree: ParamSpecTree) -> ParamNameTree:
+        """
+        Helper function to convert a ParamSpecTree-kind of tree where all
+        ParamSpecBases are substituted with their ``.name`` s.
+        Will turn {A: (B, C)} into {A.name: (B.name, C.name)}
+        """
+        name_tree: ParamNameTree = {}
+        for ps, ps_tuple in tree.items():
+            name_tree[ps.name] = tuple(p.name for p in ps_tuple)
+        return name_tree
+
+    @staticmethod
+    def _invert_tree(tree: ParamSpecTree) -> ParamSpecTree:
+        """
+        Helper function to invert a ParamSpecTree. Will turn {A: (B, C)} into
+        {B: (A,), C: (A,)}
+        """
+        indeps: Set[ParamSpecBase] = set()
+        for indep_tup in tree.values():
+            indeps.update(indep_tup)
+
+        inverted: ParamSpecTree = {}
+        for indep in indeps:
+            deps = tuple(ps for ps in tree if indep in tree[ps])
+            inverted[indep] = deps
+
+        return inverted
+
+    @staticmethod
+    def _remove_duplicates(tree: ParamSpecTree) -> None:
+        """
+        Helper function to remove duplicate entries from a ParamSpecTree while
+        preserving order. Will turn {A: (B, B, C)} into {A: (B, C)}
+        """
+        for ps, tup in tree.items():
+            specs: List[ParamSpecBase] = []
+            for p in tup:
+                if p not in specs:
+                    specs.append(p)
+            tree[ps] = tuple(specs)
+
+    @staticmethod
+    def _id(ps: ParamSpecBase) -> str:
```

Comment:
as we discussed earlier, we need to get rid of this and use `name` everywhere instead. There's not added value in `id` so far and `name` has to be unique within interdeps object.

Reply:
Okay.